var/spack/repos/builtin/packages/dyninst/package.py

```diff
@@ -20,78 +20,99 @@ class Dyninst(Package):
     patch('stackanalysis_h.patch', when='@9.2.0')
 
-    # new version uses cmake
-    def install(self, spec, prefix):
-        if spec.satisfies('@:8.1'):
-            configure("--prefix=" + prefix)
-            make()
-            make("install")
-            return
-
-        libelf = spec['elf'].prefix
-        if spec.satisfies('@:9'):
-            libdwarf = spec['libdwarf'].prefix
-
-        with working_dir('spack-build', create=True):
-            args = ['..',
-                    '-DBoost_INCLUDE_DIR=%s'    % spec['boost'].prefix.include,
-                    '-DBoost_LIBRARY_DIR=%s'    % spec['boost'].prefix.lib,
-                    '-DBoost_NO_SYSTEM_PATHS=TRUE',
-                    '-DLIBELF_INCLUDE_DIR=%s'   % join_path(
-                        libelf.include, 'libelf'),
-                    '-DLIBELF_LIBRARIES=%s'     % join_path(
-                        libelf.lib, "libelf." + dso_suffix)]
-            if spec.satisfies('@:9'):
-                args.append('-DLIBDWARF_INCLUDE_DIR=%s' % libdwarf.include)
-                args.append('-DLIBDWARF_LIBRARIES=%s'   % join_path(
-                    libdwarf.lib, "libdwarf." + dso_suffix))
-            # For @develop + use elfutils libdw, libelf is an abstraction
-            # we are really using elfutils here
-            if spec.satisfies('@develop') or spec.satisfies('@10:'):
-                tbb = spec['tbb'].prefix
-                args.append('-DTBB_INCLUDE_DIRS=%s' % tbb.include)
-                args.append('-DTBB_LIBRARIES=%s'    % join_path(
-                    tbb.lib, "libtbb." + dso_suffix))
-                args.append('-DLIBDWARF_INCLUDE_DIR=%s' % libelf.include)
-                args.append('-DLIBDWARF_LIBRARIES=%s'   % join_path(
-                    libelf.lib, "libdw." + dso_suffix))
-            if spec.satisfies('arch=linux-redhat7-ppc64le'):
-                args.append('-Darch_ppc64_little_endian=1')
-            args += std_cmake_args
-            cmake(*args)
-            make()
-            make("install")
-
-    @when('@:8.1')
-    def install(self, spec, prefix):
-        configure("--prefix=" + prefix)
-        make()
-        make("install")
+    # Versions 9.3.x used cotire, but have no knob to turn it off.
+    # Cotire has no real use for one-time builds and can break
+    # parallel builds with both static and shared libs.
+    @when('@9.3.0:9.3.99')
+    def patch(self):
+        filter_file('USE_COTIRE true', 'USE_COTIRE false',
+                    'cmake/shared.cmake')
+
+    def cmake_args(self):
+        spec = self.spec
+
+        # Elf -- the directory containing libelf.h, and the path to
+        # libelf.so.  Elfutils puts the headers in include, but libelf
+        # uses include/libelf.
+        elf = spec['elf'].prefix
+        if spec.satisfies('@9.3.0:'):
+            elf_include = elf.include
+        else:
+            elf_include = join_path(elf.include, 'libelf')
+        elf_lib = join_path(elf.lib, 'libelf.' + dso_suffix)
```

Comment:
This logic and the logic below might be simpler if you use `spec['elf'].libs` and `spec['elf'].headers` instead of manually searching for the libraries/headers yourself. You may need to override the default `libs` and `headers` properties in libelf/elfutils but then any package that needs this info can ask `elf` where to find it instead of digging around itself.

Reply:
I looked into find_libraries().  That makes it a bit cleaner to find
libelf.so and libdw.so from elfutils.  It doesn't really reduce the
number of cases, but at least I can get rid of dso_suffix (which I
didn't like).

But the libelf vs elfutils headers are more complicated.  It looks
like spec['elf'].headers returns a list of .h files, but what I need
is the directory containing them.  I could take dirname() of one
element in the list, but that seems a bit clunky.

I don't know if there's a standard way of using the libelf headers.
That is, do you #include "libelf.h" or "libelf/libelf.h" and how you
set -I accordingly.  Libdwarf is clever enough to test what it's given
and adapt, but dyninst does not.

I didn't find anything like spec[...].headers_dir that would abstract
out the include vs include/libelf issue.  Is there one?

I'll add spec[].libs for elfutils and change a couple lines to use
find_libraries.